Improve assert_eq! and assert_ne!

a1phyr opened the pull request:
This PR improves `assert_eq!` and `assert_ne!` by moving the panicking code in an external function.

It does not change the fast path, but the move of the formatting in the cold path (the panic) may have a positive effect on in instruction cache use and with inlining.

Moreover, the use of trait objects instead of generic may improve compile times for `assert_eq!`-heavy code.

Godbolt link: ~~https://rust.godbolt.org/z/TYa9MT~~ \
Updated: https://rust.godbolt.org/z/bzE84x

a1phyr committed:
Improve assert_eq! and assert_ne!

It should improve compile times and reduce instruction cache use by moving the
panic formatting to a monomorphised function

rust-highfive commented:
r? @cramertj

(rust_highfive has picked a reviewer for you, use r? to override)

pickfire commented:
Does this improves runtime performance?

> Godbolt link: https://rust.godbolt.org/z/TYa9MT

I can't seemed to notice the difference in the assembly.

pickfire commented on library/core/src/macros/internals.rs:
```diff
@@ -0,0 +1,85 @@
+#[cold]
+#[doc(hidden)]
+#[unstable(feature = "macros_internals", reason = "macros implementation detail", issue = "none")]
+#[track_caller]
```
Do we need `#[track_caller]` here?

a1phyr replied:
I wasn't sure about this one, I put it because the `panic!` started in the inner function, and we want it to be reported near the `assert_eq!` call.

a1phyr commented:
> I can't seemed to notice the difference in the assembly.

The difference is not in the fast path, but in the panicking one.

pickfire commented:
> Moreover, the use of trait objects instead of generic may improve compile times for assert_eq!-heavy code.

Looks nice but I wonder if we need 2x/4x the trait objects? Is it possible to have one `<T, U>` (merged both `assert_eq!` and `assert_ne!` but with different output string) and maybe another one for `Arguments` (I am not sure if we can merge this too)?

- reposting as you probably missed this

pickfire commented:
> The only formatting that change between assert_eq! and assert_ne! is that (left == right) becomes (left != right). We could eventually give the comparison operator as parameter to assert_eq_failed.

Yes, but doesn't the current implementation makes it another function which doubles the code if a user use `assert_eq!` and `assert_ne!` on the same `T` and `U`?

pickfire commented:
> > I can't seemed to notice the difference in the assembly.
> 
> The difference is not in the fast path, but in the panicking one.

Diffing the two assembly outputs seemed to give me nothing, are you sure you posted the correct link? Or maybe I copied the wrong thing?

a1phyr commented:
> > The only formatting that change between `assert_eq!` and `assert_ne!` is that `(left == right)` becomes `(left != right)`. We could eventually give the comparison operator as parameter to `assert_eq_failed`.
> 
> Yes, but doesn't the current implementation makes it another function which doubles the code if a user use `assert_eq!` and `assert_ne!` on the same `T` and `U`?

Yes, it does. We can add an `operator` parameter, which is equal to `==` for `assert_eq!` and `!=` for `assert_ne!`, and use it for the formatting later. 

> Not sure about the `Arguments` case since I am not sure if that could be optimized into a separate function.

Well, it could, but we definitely want to keep a separate generic function when there is no `Argument`, because it has to be created in the function calling `assert_eq`, and we want the generated code here as thin as possible. IMHO it is not worth the trouble.

a1phyr commented:
> Diffing the two assembly outputs seemed to give me nothing, are you sure you posted the correct link? Or maybe I copied the wrong thing?

You don't have any difference between `old_basic` and `new_basic` ? Or between `old_args` and `new_args` ? Weird... I checked the link, it works on bot my computer and my phone.

andjo403 commented:
think that @pickfire expected something like this https://rust.godbolt.org/z/4co4f1 where there is a diff between two editors that implement the same thing in two ways. I think that I split @a1phyr:s example correct in that godbolt link.

a1phyr commented:
@andjo403 Oh, I see. I didn't know you could produce diffs like that in Godbold.

a1phyr committed:
Fix UI tests and merge `assert_eq` and `assert_ne` internal functions

a1phyr commented:
This is very strange, some tests fail but seem totally unrelated (eg a compile-fail test about revisions), and I'm not sure what to do about them.

Anyway, let's see if it improves compile time

@bors try @rust-timer queue

rust-timer commented:
Insufficient permissions to issue commands to rust-timer.

bors commented:
@a1phyr: :key: Insufficient privileges: not in try users

bors commented:
@a1phyr: :key: Insufficient privileges: not in try users

rust-timer commented:
Insufficient permissions to issue commands to rust-timer.

pickfire commented:
@andjo403 Oh, I didn't know you can produce diff in godbolt.

It seemed like the assembly lines increased.

pickfire commented:
Updated godbolt to include 6 functions, it seemed like the number of assembly lines increased, removing `#[track_caller]` can slightly reduce it (380 vs 512). https://rust.godbolt.org/z/EsGWbG

Merging the one with `Arguments` into the same function seemed to produce a slightly more lines of assembly (512 -> 540). https://rust.godbolt.org/z/a7rqrj

a1phyr commented:
Can someone start rust-timer to see if this has a positive impact on build time ?

And also, does someone has a clue what makes [this test](https://github.com/rust-lang/rust/blob/master/src/test/compile-fail/meta-expected-error-wrong-rev.rs) fail ? It does not seem related to my changes.

pickfire commented:
@jyn514

jyn514 commented:
@pickfire please don't ping me specifically for perf runs, there's a [Zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Perf.20run) for that.

@a1phyr are you sure it makes sense to measure the perf before fixing the tests? Your fix might itself have a perf impact.

a1phyr commented:
I think the only changes that I have left to do are tests themselves (eg changing a mir output), but you're right that we should wait them to be fixed.

jyn514 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 0dbf6d0f280da53dfadc453a699d01b577a33a4b with merge 7827eb84fe5a82fab0dcc16f6eb5121188f2846e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0dbf6d0f280da53dfadc453a699d01b577a33a4b","merge_sha":"7827eb84fe5a82fab0dcc16f6eb5121188f2846e"} -->

m-ou-se commented:
>    * This make `assert_eq!` always call `core::panic!`.
> 
>    * `assert_eq!` and `assert_ne!` don't require `panic!` to be in scope anymore.
> 
>    * **`panic!` cannot be overridden by users of `assert_eq!` anymore**. Was this part of our stability guaranties ?

There's already an open PR that does these things: https://github.com/rust-lang/rust/pull/78343

Note that moving the panics these asserts produce into a function will break Clippy, so would also require changes there.

---

There's also another PR that's making changing to these two macros: https://github.com/rust-lang/rust/pull/79001

a1phyr commented:
Thanks for the links ! I updated the top-level comment to include the first one.

About Clippy, we'll just have to wait for #79145 to be merged, I guess.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1420075233)
Build commit: 7827eb84fe5a82fab0dcc16f6eb5121188f2846e (`7827eb84fe5a82fab0dcc16f6eb5121188f2846e`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1420075233"},"merge_sha":"7827eb84fe5a82fab0dcc16f6eb5121188f2846e"} -->

rust-timer commented:
Queued 7827eb84fe5a82fab0dcc16f6eb5121188f2846e with parent 7d747db0d5dd8f08f2efb073e2e77a34553465a7, future [comparison URL](https://perf.rust-lang.org/compare.html?start=7d747db0d5dd8f08f2efb073e2e77a34553465a7&end=7827eb84fe5a82fab0dcc16f6eb5121188f2846e).

m-ou-se commented:
That fix for Clippy only works if `assert_eq`/`assert_ne` expand to a call to one of those panic functions. Clippy will not look *inside* any functions that the macro expands to. So you'll at least have to add these functions to the list of functions Clippy recognizes as a panic.

a1phyr commented:
Oh, I see.

Should I do it in a different PR ? It think not because we don't want to change clippy if this is not merged

rust-timer commented:
Finished benchmarking try commit (7827eb84fe5a82fab0dcc16f6eb5121188f2846e): [comparison url](https://perf.rust-lang.org/compare.html?start=7d747db0d5dd8f08f2efb073e2e77a34553465a7&end=7827eb84fe5a82fab0dcc16f6eb5121188f2846e).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-perf

pickfire commented:
@a1phyr Want to try using the other method that I showed? Merging all of them into one function.

JohnCSimon commented:
@a1phyr 
Ping from triage - can you please address the pickfire's comments and resolve the merge conflicts? Thank you.

pickfire commented:
@JohnCSimon If we want to try out a benchmark using a different method should we create a new pull request and bors try to compare?

Dylan-DPC commented:
@pickfire yeah or wait for this to get merged and then make another PR and compare it to master

a1phyr committed:
Apply suggestion

a1phyr commented:
I have just rebased my branch on `master` and applied the suggestion to merge `assert_failed` and `assert_failed_args`.

I didn't exactly used @pickfire's code because it uses a `String` which is not available in `core`. I tried to use another `fmt::Argument` instead but I didn't manage to get rid of lifetime issues, so I kept the duplication of the format string.

m-ou-se commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 52f686005a39df287f3dda0b8776a7a091dda8e4 with merge ba6e65067e5bf513aed949a29570f183e9b6e441...
<!-- homu: {"type":"TryBuildStarted","head_sha":"52f686005a39df287f3dda0b8776a7a091dda8e4","merge_sha":"ba6e65067e5bf513aed949a29570f183e9b6e441"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1523573562)
Build commit: ba6e65067e5bf513aed949a29570f183e9b6e441 (`ba6e65067e5bf513aed949a29570f183e9b6e441`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1523573562"},"merge_sha":"ba6e65067e5bf513aed949a29570f183e9b6e441"} -->

rust-timer commented:
Queued ba6e65067e5bf513aed949a29570f183e9b6e441 with parent db85512bd8d922e0785029eff41152b0f35690f1, future [comparison URL](https://perf.rust-lang.org/compare.html?start=db85512bd8d922e0785029eff41152b0f35690f1&end=ba6e65067e5bf513aed949a29570f183e9b6e441).

rust-timer commented:
Finished benchmarking try commit (ba6e65067e5bf513aed949a29570f183e9b6e441): [comparison url](https://perf.rust-lang.org/compare.html?start=db85512bd8d922e0785029eff41152b0f35690f1&end=ba6e65067e5bf513aed949a29570f183e9b6e441).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-perf

m-ou-se commented:
> This PR does the following changes (...)

Note that that's no longer changed by this PR. https://github.com/rust-lang/rust/pull/78343 already changed that.

a1phyr commented:
> Note that that's no longer changed by this PR. #78343 already changed that.

You're right, it is removed

m-ou-se commented:
> I wasn't sure where to put internals functions, so I put them in macros/internals.rs, but there may be a better solution.

`library/core/src/panicking.rs` might be a good place for these functions. That's also where the internal functions of `panic!()` are.

sivadeilra commented:
Could I make a suggestion?  Instead of passing string literals for `"=="` and `"!="`, define a very small `enum` for these and pass the enum value instead of the string literal. All of the functions that use `assert_eq` and `assert_ne` will need to contain a basic block that calls the failure path of course, and if you use an enum instead of a string literal, there are several benefits:
1. In most calling conventions, you'll use only 1 argument register (for the enum) instead of 2 registers (str data pointer and str len). This leaves more registers for other arguments, and may lead to less code generated for register spilling / calling.
2. you won't pay the cost of locating a global static item. With PIC code, usually there is an `leaq` instruction or something similar which computes the absolute address of a PIC-relative static item. With non-PIC but relocatable code, there's a relocation record.

These might seem like small benefits, but if there are many uses of `assert_ne` and such, then it can add up.

a1phyr committed:
Apply suggestions

- Move `assert_failed` to core::panicking`
- Make `assert_failed` use an enum instead of a string

rust-log-analyzer commented:
The job `mingw-check` of your PR [failed](https://github.com/rust-lang/rust/runs/1566509809) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang/rust/commit/00b0bfa1cf4e38774e44ae2ba303717e5a356125/checks/1566509809/logs), [raw log](https://github.com/rust-lang/rust/commit/00b0bfa1cf4e38774e44ae2ba303717e5a356125/checks/1566509809/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[command]/usr/bin/git submodule foreach --recursive git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :
[command]/usr/bin/git config --local http.https://github.com/.extraheader AUTHORIZATION: basic ***
##[endgroup]
##[group]Fetching the repository
[command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=2 origin +2bbfe02166849b1071afc4a98fe12d6ac8d03762:refs/remotes/pull/79100/merge
---
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/core/src/panicking.rs"` failed.
Diff in /checkout/library/core/src/panicking.rs at line 94:
 
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
 #[derive(Debug)]
 pub enum AssertKind {
-    Eq, Ne
+    Eq,
+    Ne,
 
 
 /// Internal function for `assert_eq!` and `assert_ne!` macros
Diff in /checkout/library/core/src/panicking.rs at line 101:
 #[cold]
 #[track_caller]
-pub fn assert_failed<T, U>(kind: AssertKind, left: &T, right: &U, args: Option<fmt::Arguments<'_>>) -> !
+pub fn assert_failed<T, U>(
+    kind: AssertKind,
+    left: &T,
+    right: &U,
+    args: Option<fmt::Arguments<'_>>,
+) -> !
 where
     T: fmt::Debug + ?Sized,
     U: fmt::Debug + ?Sized,
Build completed unsuccessfully in 0:00:19
== clock drift check ==
  local time: Wed Dec 16 21:18:18 UTC 2020
  network time: Wed, 16 Dec 2020 10:06:49 GMT
```

</details><p></p>

[I'm a bot](https://github.com/rust-lang/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-lang/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

JohnCSimon commented:
Ping from triage
@a1phyr - what is the status of this PR? Is it ready for review? Thank you!

a1phyr commented:
This PR is ready for review

bors commented:
:umbrella: The latest upstream changes (presumably #79328) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

a1phyr commented:
Rebased

jyn514 commented on library/core/src/panicking.rs:
```diff
@@ -91,3 +91,52 @@ pub fn panic_fmt(fmt: fmt::Arguments<'_>) -> ! {
 }
+
+#[derive(Debug)]
+pub enum AssertKind {
```
```suggestion
#[doc(hidden)]
pub enum AssertKind {
```

jyn514 commented on library/core/src/panicking.rs:
```diff
@@ -91,3 +91,52 @@ pub fn panic_fmt(fmt: fmt::Arguments<'_>) -> ! {
+/// Internal function for `assert_eq!` and `assert_ne!` macros
+#[cold]
+#[track_caller]
+pub fn assert_failed<T, U>(
```
```suggestion
#[doc(hidden)]
pub fn assert_failed<T, U>(
```

pickfire commented:
Do we need to do a perf run again since two functions are changed into one?

a1phyr commented:
> Do we need to do a perf run again since two functions are changed into one with the enum?

The last perf run was made with the two function merged. We can do another one to see the effect of using an enum, but I can't see how it could be negative.

a1phyr committed:
Hide internals items in documentation

Co-authored-by: Joshua Nelson <joshua@yottadb.com>

bors commented:
:umbrella: The latest upstream changes (presumably #80851) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

a1phyr commented:
Rebased

a1phyr commented:
Does this PR need anything to be merged ?

m-ou-se commented:
> Does this PR need anything to be merged ?

Yes, it needs approval. :)

@bors r+ rollup=never

Thanks for working on this!

bors commented:
:pushpin: Commit b973970e3dccf0b2f072f1956f2d563d323b4cff has been approved by `m-ou-se`

<!-- @bors r=m-ou-se b973970e3dccf0b2f072f1956f2d563d323b4cff -->
<!-- homu: {"type":"Approved","sha":"b973970e3dccf0b2f072f1956f2d563d323b4cff","approver":"m-ou-se"} -->

bors commented:
:hourglass: Testing commit b973970e3dccf0b2f072f1956f2d563d323b4cff with merge 92b84c9a0f30fa137ac2bce7a941fec0b3691849...
<!-- homu: {"type":"BuildStarted","head_sha":"b973970e3dccf0b2f072f1956f2d563d323b4cff","merge_sha":"92b84c9a0f30fa137ac2bce7a941fec0b3691849"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1836446121) [(plain)](https://github.com/rust-lang-ci/rust/commit/92b84c9a0f30fa137ac2bce7a941fec0b3691849/checks/1836446121/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
status: exit code: 2
command: "make"
stdout:
------------------------------------------
make[1]: Entering directory '/d/a/rust/rust/src/test/run-make-fulldeps/coverage-spanview'
# Compile the test library with coverage instrumentation
PATH="/d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview:D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2\bin:/c/Program Files (x86)/Windows Kits/10/bin/x64:/c/Program Files (x86)/Windows Kits/10/bin/10.0.19041.0/x64:/c/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.28.29333/bin/HostX64/x64:/c/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Tools/MSVC/14.28.29333/bin/HostX64/x64:/d/a/rust/rust/build/x86_64-pc-windows-msvc/stage0-bootstrap-tools/x86_64-pc-windows-msvc/release/deps:/d/a/rust/rust/build/x86_64-pc-windows-msvc/stage0/bin:/d/a/rust/rust/ninja:/d/a/rust/rust/msys2/mingw64/bin:/c/hostedtoolcache/windows/Python/3.9.1/x64/Scripts:/c/hostedtoolcache/windows/Python/3.9.1/x64:/usr/bin:/d/a/rust/rust/sccache:/c/Users/runneradmin/.dotnet/tools:/c/Program Files/MongoDB/Server/4.4/bin:/c/aliyun-cli:/c/vcpkg:/c/cf-cli:/c/Program Files (x86)/NSIS:/c/Program Files/Mercurial:/c/hostedtoolcache/windows/stack/2.5.1/x64:/c/ProgramData/chocolatey/lib/ghc.8.10.3/tools/ghc-8.10.3/bin:/c/Program Files/dotnet:/c/mysql-5.7.21-winx64/bin:/c/Program Files/R/R-4.0.3/bin/x64:/c/SeleniumWebDrivers/GeckoDriver:/c/Program Files (x86)/sbt/bin:/c/Rust/.cargo/bin:/c/Program Files (x86)/GitHub CLI:/c/Program Files/Git/bin:/c/Program Files (x86)/pipx_bin:/c/hostedtoolcache/windows/go/1.14.14/x64/bin:/c/hostedtoolcache/windows/Python/3.7.9/x64/Scripts:/c/hostedtoolcache/windows/Python/3.7.9/x64:/c/hostedtoolcache/windows/Ruby/2.5.8/x64/bin:/c/Program Files/Java/jdk8u282-b08/bin:/c/npm/prefix:/c/Program Files/Microsoft SDKs/Azure/Azure Dev Spaces CLI:/c/Program Files/Microsoft SDKs/Azure/Azure Dev Spaces CLI:/c/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin:/c/ProgramData/kind:/c/windows/system32:/c/windows:/c/windows/System32/Wbem:/c/windows/System32/WindowsPowerShell/v1.0:/c/windows/System32/OpenSSH:/c/ProgramData/Chocolatey/bin:/c/Program Files/Microsoft/Web Platform Installer:/c/Program Files/Docker:/c/Program Files/PowerShell/7:/c/Program Files/dotnet:/c/Program Files/Microsoft SQL Server/130/Tools/Binn:/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn:/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit:/c/Program Files (x86)/Microsoft SQL Server/110/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/120/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/130/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/140/DTS/Binn:/c/Program Files (x86)/Microsoft SQL Server/150/DTS/Binn:/c/Program Files/nodejs:/c/ProgramData/chocolatey/lib/pulumi/tools/Pulumi/bin:/c/ProgramData/chocolatey/lib/maven/apache-maven-3.6.3/bin:/c/Program Files/Microsoft Service Fabric/bin/Fabric/Fabric.Code:/c/Program Files/Microsoft SDKs/Service Fabric/Tools/ServiceFabricLocalClusterManager:/c/Program Files/OpenSSL/bin:/c/Strawberry/c/bin:/c/Strawberry/perl/site/bin:/c/Strawberry/perl/bin:/c/Program Files/Git/cmd:/c/Program Files/Git/mingw64/bin:/c/Program Files/Git/usr/bin:/c/tools/php:/c/Program Files (x86)/sbt/bin:/c/Program Files/TortoiseSVN/bin:/c/SeleniumWebDrivers/ChromeDriver:/c/SeleniumWebDrivers/EdgeDriver:/c/Program Files/CMake/bin:/c/Program Files/Amazon/AWSCLIV2:/c/Program Files/Amazon/SessionManagerPlugin/bin:/c/Program Files/Amazon/AWSSAMCLI/bin:/c/Program Files (x86)/Google/Cloud SDK/google-cloud-sdk/bin:/c/Program Files (x86)/Microsoft BizTalk Server:/c/Users/runneradmin/AppData/Local/Microsoft/WindowsApps" 'D:\a\rust\rust\build\x86_64-pc-windows-msvc\stage2\bin\rustc.exe' --out-dir /d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview -L /d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview  ../coverage/lib/doctest_crate.rs \
  $( grep -q '^\/\/ require-rust-edition-2018' ../coverage/lib/doctest_crate.rs && echo "--edition=2018" ) \
Some tests failed in compiletest suite=run-make-fulldeps mode=run-make host=x86_64-pc-windows-msvc target=x86_64-pc-windows-msvc
  -Zinstrument-coverage \
  -Zinstrument-coverage \
  -Zdump-mir=InstrumentCoverage \
  -Zdump-mir-spanview \
  -Zdump-mir-dir="/d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview"/mir_dump.doctest_crate
for path in "/d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview"/mir_dump.doctest_crate/*; do \
 file="$(basename "$path")"; \
 urlescaped="$("C:\hostedtoolcache\windows\Python\3.9.1\x64\python3.exe" ../coverage-spanview/escape_url.py $file)" || exit $?; \
 printf "$SPANVIEW_HEADER\n" "doctest_crate" "$urlescaped" > "$path".modified; \
 tail -n +2 "$path" >> "$path".modified; \
 mv "$path".modified "$path"; \
done && true # for/done ends in non-zero status
# Check that the selected `mir_dump` files match what we expect (`--bless` refreshes `expected`)
mkdir -p "/d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview"/actual_mir_dump.doctest_crate
rm -f "/d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview"/actual_mir_dump.doctest_crate/*
cp "/d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview"/mir_dump.doctest_crate/*InstrumentCoverage.0.html "/d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview"/actual_mir_dump.doctest_crate/
diff -u --strip-trailing-cr -r expected_mir_dump.doctest_crate/ "/d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview"/actual_mir_dump.doctest_crate/
diff -u --strip-trailing-cr -r expected_mir_dump.doctest_crate/doctest_crate.fn_run_in_doctests.-------.InstrumentCoverage.0.html /d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview/actual_mir_dump.doctest_crate/doctest_crate.fn_run_in_doctests.-------.InstrumentCoverage.0.html
--- expected_mir_dump.doctest_crate/doctest_crate.fn_run_in_doctests.-------.InstrumentCoverage.0.html 2021-02-05 05:11:33.889006600 +0000
+++ /d/a/rust/rust/build/x86_64-pc-windows-msvc/test/run-make-fulldeps/coverage-spanview/coverage-spanview/actual_mir_dump.doctest_crate/doctest_crate.fn_run_in_doctests.-------.InstrumentCoverage.0.html 2021-02-05 06:10:06.568753300 +0000
@@ -71,103 +71,43 @@
 <body>
 <div class="code" style="counter-reset: line 1"><span class="line"><span><span class="code even" style="--layer: 1"><span class="annotation">@0⦊</span>pub fn fn_run_in_doctests(conditional: usize) <span class="annotation">⦉@0</span></span></span><span class="code" style="--layer: 0">{</span></span>
 <span class="line"><span class="code" style="--layer: 0">    match </span><span><span class="code even" style="--layer: 1" title="3:11-3:22: @0[0]: FakeRead(ForMatchedPlace, _1)"><span class="annotation">@0⦊</span>conditional<span class="annotation">⦉@0</span></span></span><span class="code" style="--layer: 0"> {</span></span>
-<span class="line"><span class="code" style="--layer: 0">        1 =&gt; </span><span><span class="code odd" style="--layer: 1" title="4:14-4:30: @7[0]: _0 = const ()"><span class="annotation">@7⦊</span></span></span><span class="code even" style="--layer: 2" title="4:14-4:30: @6[5]: _138 = const fn_run_in_doctests::promoted[0]
-4:14-4:30: @6[6]: _17 = &amp;(*_138)
-4:14-4:30: @6[7]: _16 = &amp;(*_17)
-4:14-4:30: @6[8]: _15 = move _16 as &amp;[&amp;str] (Pointer(Unsize))
-4:14-4:30: @6[17]: _25 = &amp;(*_7)
-4:14-4:30: @6[18]: _24 = &amp;_25
-4:14-4:30: @6[21]: _27 = &amp;(*_8)
-4:14-4:30: @6[22]: _26 = &amp;_27
-4:14-4:30: @6[23]: _23 = (move _24, move _26)
-4:14-4:30: @6[26]: FakeRead(ForMatchedPlace, _23)
-4:14-4:30: @6[28]: _28 = (_23.0: &amp;&amp;i32)
-4:14-4:30: @6[30]: _29 = (_23.1: &amp;&amp;i32)
-4:14-4:30: @6[33]: _31 = &amp;(*_28)
-4:14-4:30: @6[35]: _32 = &lt;&amp;i32 as Debug&gt;::fmt as for&lt;&#39;r, &#39;s, &#39;t0&gt; fn(&amp;&#39;r &amp;i32, &amp;&#39;s mut std::fmt::Formatter&lt;&#39;t0&gt;) -&gt; std::result::Result&lt;(), std::fmt::Error&gt; (Pointer(ReifyFnPointer))
-4:14-4:30: @6.Call: _30 = ArgumentV1::new::&lt;&amp;i32&gt;(move _31, move _32) -&gt; [return: bb8, unwind: bb29]
-4:14-4:30: @8[4]: _34 = &amp;(*_29)
-4:14-4:30: @8[6]: _35 = &lt;&amp;i32 as Debug&gt;::fmt as for&lt;&#39;r, &#39;s, &#39;t0&gt; fn(&amp;&#39;r &amp;i32, &amp;&#39;s mut std::fmt::Formatter&lt;&#39;t0&gt;) -&gt; std::result::Result&lt;(), std::fmt::Error&gt; (Pointer(ReifyFnPointer))
-4:14-4:30: @8.Call: _33 = ArgumentV1::new::&lt;&amp;i32&gt;(move _34, move _35) -&gt; [return: bb9, unwind: bb29]
-4:14-4:30: @9[2]: _22 = [move _30, move _33]
-4:14-4:30: @9[7]: _21 = &amp;_22
-4:14-4:30: @9[8]: _20 = &amp;(*_21)
-4:14-4:30: @9[9]: _19 = move _20 as &amp;[std::fmt::ArgumentV1] (Pointer(Unsize))
-4:14-4:30: @9.Call: _14 = Arguments::new_v1(move _15, move _19) -&gt; [return: bb10, unwind: bb29]
-4:14-4:30: @10.Call: core::panicking::panic_fmt(move _14) -&gt; bb29"><span class="annotation">@6,8,9,10⦊</span>assert_eq!(1, 1)<span class="annotation">⦉@6,8,9,10</span></span><span><span class="code odd" style="--layer: 1" title="4:14-4:30: @7[0]: _0 = const ()"><span class="annotation">⦉@7</span></span></span><span class="code" style="--layer: 0">, // this is run,</span></span>
-<span class="line"><span class="code" style="--layer: 0">        2 =&gt; </span><span><span class="code even" style="--layer: 1" title="5:14-5:30: @13[0]: _0 = const ()"><span class="annotation">@13⦊</span></span></span><span class="code even" style="--layer: 2" title="5:14-5:30: @12[5]: _141 = const fn_run_in_doctests::promoted[3]
-5:14-5:30: @12[6]: _51 = &amp;(*_141)
-5:14-5:30: @12[7]: _50 = &amp;(*_51)
-5:14-5:30: @12[8]: _49 = move _50 as &amp;[&amp;str] (Pointer(Unsize))
-5:14-5:30: @12[17]: _59 = &amp;(*_41)
-5:14-5:30: @12[18]: _58 = &amp;_59
-5:14-5:30: @12[21]: _61 = &amp;(*_42)
-5:14-5:30: @12[22]: _60 = &amp;_61
-5:14-5:30: @12[23]: _57 = (move _58, move _60)
-5:14-5:30: @12[26]: FakeRead(ForMatchedPlace, _57)
-5:14-5:30: @12[28]: _62 = (_57.0: &amp;&amp;i32)
-5:14-5:30: @12[30]: _63 = (_57.1: &amp;&amp;i32)
-5:14-5:30: @12[33]: _65 = &amp;(*_62)
-5:14-5:30: @12[35]: _66 = &lt;&amp;i32 as Debug&gt;::fmt as for&lt;&#39;r, &#39;s, &#39;t0&gt; fn(&amp;&#39;r &amp;i32, &amp;&#39;s mut std::fmt::Formatter&lt;&#39;t0&gt;) -&gt; std::result::Result&lt;(), std::fmt::Error&gt; (Pointer(ReifyFnPointer))
-5:14-5:30: @12.Call: _64 = ArgumentV1::new::&lt;&amp;i32&gt;(move _65, move _66) -&gt; [return: bb14, unwind: bb29]
-5:14-5:30: @14[4]: _68 = &amp;(*_63)
-5:14-5:30: @14[6]: _69 = &lt;&amp;i32 as Debug&gt;::fmt as for&lt;&#39;r, &#39;s, &#39;t0&gt; fn(&amp;&#39;r &amp;i32, &amp;&#39;s mut std::fmt::Formatter&lt;&#39;t0&gt;) -&gt; std::result::Result&lt;(), std::fmt::Error&gt; (Pointer(ReifyFnPointer))
-5:14-5:30: @14.Call: _67 = ArgumentV1::new::&lt;&amp;i32&gt;(move _68, move _69) -&gt; [return: bb15, unwind: bb29]
-5:14-5:30: @15[2]: _56 = [move _64, move _67]
-5:14-5:30: @15[7]: _55 = &amp;_56
-5:14-5:30: @15[8]: _54 = &amp;(*_55)
-5:14-5:30: @15[9]: _53 = move _54 as &amp;[std::fmt::ArgumentV1] (Pointer(Unsize))
-5:14-5:30: @15.Call: _48 = Arguments::new_v1(move _49, move _53) -&gt; [return: bb16, unwind: bb29]
-5:14-5:30: @16.Call: core::panicking::panic_fmt(move _48) -&gt; bb29"><span class="annotation">@12,14,15,16⦊</span>assert_eq!(1, 1)<span class="annotation">⦉@12,14,15,16</span></span><span><span class="code even" style="--layer: 1" title="5:14-5:30: @13[0]: _0 = const ()"><span class="annotation">⦉@13</span></span></span><span class="code" style="--layer: 0">, // this,</span></span>
-<span class="line"><span class="code" style="--layer: 0">        3 =&gt; </span><span><span class="code odd" style="--layer: 1" title="6:14-6:30: @19[0]: _0 = const ()"><span class="annotation">@19⦊</span></span></span><span class="code even" style="--layer: 2" title="6:14-6:30: @18[5]: _144 = const fn_run_in_doctests::promoted[6]
-6:14-6:30: @18[6]: _85 = &amp;(*_144)
-6:14-6:30: @18[7]: _84 = &amp;(*_85)
-6:14-6:30: @18[8]: _83 = move _84 as &amp;[&amp;str] (Pointer(Unsize))
-6:14-6:30: @18[17]: _93 = &amp;(*_75)
-6:14-6:30: @18[18]: _92 = &amp;_93
-6:14-6:30: @18[21]: _95 = &amp;(*_76)
-6:14-6:30: @18[22]: _94 = &amp;_95
-6:14-6:30: @18[23]: _91 = (move _92, move _94)
-6:14-6:30: @18[26]: FakeRead(ForMatchedPlace, _91)
-6:14-6:30: @18[28]: _96 = (_91.0: &amp;&amp;i32)
-6:14-6:30: @18[30]: _97 = (_91.1: &amp;&amp;i32)
-6:14-6:30: @18[33]: _99 = &amp;(*_96)
-6:14-6:30: @18[35]: _100 = &lt;&amp;i32 as Debug&gt;::fmt as for&lt;&#39;r, &#39;s, &#39;t0&gt; fn(&amp;&#39;r &amp;i32, &amp;&#39;s mut std::fmt::Formatter&lt;&#39;t0&gt;) -&gt; std::result::Result&lt;(), std::fmt::Error&gt; (Pointer(ReifyFnPointer))
-6:14-6:30: @18.Call: _98 = ArgumentV1::new::&lt;&amp;i32&gt;(move _99, move _100) -&gt; [return: bb20, unwind: bb29]
-6:14-6:30: @20[4]: _102 = &amp;(*_97)
-6:14-6:30: @20[6]: _103 = &lt;&amp;i32 as Debug&gt;::fmt as for&lt;&#39;r, &#39;s, &#39;t0&gt; fn(&amp;&#39;r &amp;i32, &amp;&#39;s mut std::fmt::Formatter&lt;&#39;t0&gt;) -&gt; std::result::Result&lt;(), std::fmt::Error&gt; (Pointer(ReifyFnPointer))
-6:14-6:30: @20.Call: _101 = ArgumentV1::new::&lt;&amp;i32&gt;(move _102, move _103) -&gt; [return: bb21, unwind: bb29]
-6:14-6:30: @21[2]: _90 = [move _98, move _101]
-6:14-6:30: @21[7]: _89 = &amp;_90
-6:14-6:30: @21[8]: _88 = &amp;(*_89)
-6:14-6:30: @21[9]: _87 = move _88 as &amp;[std::fmt::ArgumentV1] (Pointer(Unsize))
-6:14-6:30: @21.Call: _82 = Arguments::new_v1(move _83, move _87) -&gt; [return: bb22, unwind: bb29]
-6:14-6:30: @22.Call: core::panicking::panic_fmt(move _82) -&gt; bb29"><span class="annotation">@18,20,21,22⦊</span>assert_eq!(1, 1)<span class="annotation">⦉@18,20,21,22</span></span><span><span class="code odd" style="--layer: 1" title="6:14-6:30: @19[0]: _0 = const ()"><span class="annotation">⦉@19</span></span></span><span class="code" style="--layer: 0">, // and this too</span></span>
-<span class="line"><span class="code" style="--layer: 0">        _ =&gt; </span><span><span class="code even" style="--layer: 1" title="7:14-7:30: @24[0]: _0 = const ()"><span class="annotation">@24⦊</span></span></span><span class="code even" style="--layer: 2" title="7:14-7:30: @23[5]: _147 = const fn_run_in_doctests::promoted[9]
-7:14-7:30: @23[6]: _119 = &amp;(*_147)
-7:14-7:30: @23[7]: _118 = &amp;(*_119)
-7:14-7:30: @23[8]: _117 = move _118 as &amp;[&amp;str] (Pointer(Unsize))
-7:14-7:30: @23[17]: _127 = &amp;(*_109)
-7:14-7:30: @23[18]: _126 = &amp;_127
-7:14-7:30: @23[21]: _129 = &amp;(*_110)
-7:14-7:30: @23[22]: _128 = &amp;_129
-7:14-7:30: @23[23]: _125 = (move _126, move _128)
-7:14-7:30: @23[26]: FakeRead(ForMatchedPlace, _125)
-7:14-7:30: @23[28]: _130 = (_125.0: &amp;&amp;i32)
-7:14-7:30: @23[30]: _131 = (_125.1: &amp;&amp;i32)
-7:14-7:30: @23[33]: _133 = &amp;(*_130)
-7:14-7:30: @23[35]: _134 = &lt;&amp;i32 as Debug&gt;::fmt as for&lt;&#39;r, &#39;s, &#39;t0&gt; fn(&amp;&#39;r &amp;i32, &amp;&#39;s mut std::fmt::Formatter&lt;&#39;t0&gt;) -&gt; std::result::Result&lt;(), std::fmt::Error&gt; (Pointer(ReifyFnPointer))
-7:14-7:30: @23.Call: _132 = ArgumentV1::new::&lt;&amp;i32&gt;(move _133, move _134) -&gt; [return: bb25, unwind: bb29]
-7:14-7:30: @25[4]: _136 = &amp;(*_131)
-7:14-7:30: @25[6]: _137 = &lt;&amp;i32 as Debug&gt;::fmt as for&lt;&#39;r, &#39;s, &#39;t0&gt; fn(&amp;&#39;r &amp;i32, &amp;&#39;s mut std::fmt::Formatter&lt;&#39;t0&gt;) -&gt; std::result::Result&lt;(), std::fmt::Error&gt; (Pointer(ReifyFnPointer))
-7:14-7:30: @25.Call: _135 = ArgumentV1::new::&lt;&amp;i32&gt;(move _136, move _137) -&gt; [return: bb26, unwind: bb29]
-7:14-7:30: @26[2]: _124 = [move _132, move _135]
-7:14-7:30: @26[7]: _123 = &amp;_124
-7:14-7:30: @26[8]: _122 = &amp;(*_123)
-7:14-7:30: @26[9]: _121 = move _122 as &amp;[std::fmt::ArgumentV1] (Pointer(Unsize))
-7:14-7:30: @26.Call: _116 = Arguments::new_v1(move _117, move _121) -&gt; [return: bb27, unwind: bb29]
-7:14-7:30: @27.Call: core::panicking::panic_fmt(move _116) -&gt; bb29"><span class="annotation">@23,25,26,27⦊</span>assert_eq!(1, 2)<span class="annotation">⦉@23,25,26,27</span></span><span><span class="code even" style="--layer: 1" title="7:14-7:30: @24[0]: _0 = const ()"><span class="annotation">⦉@24</span></span></span><span class="code" style="--layer: 0">, // however this is not</span></span>
+<span class="line"><span class="code" style="--layer: 0">        1 =&gt; </span><span><span class="code odd" style="--layer: 1" title="4:14-4:30: @7[0]: _0 = const ()"><span class="annotation">@7⦊</span></span></span><span class="code even" style="--layer: 2" title="4:14-4:30: @6[1]: _14 = core::panicking::AssertKind::Eq
+4:14-4:30: @6[2]: FakeRead(ForLet, _14)
+4:14-4:30: @6[5]: _16 = move _14
+4:14-4:30: @6[8]: _18 = &amp;(*_7)
+4:14-4:30: @6[9]: _17 = &amp;(*_18)
+4:14-4:30: @6[12]: _20 = &amp;(*_8)
+4:14-4:30: @6[13]: _19 = &amp;(*_20)
+4:14-4:30: @6[15]: _21 = Option::&lt;Arguments&gt;::None
+4:14-4:30: @6.Call: core::panicking::assert_failed::&lt;i32, i32&gt;(move _16, move _17, move _19, move _21) -&gt; bb17"><span class="annotation">@6⦊</span>assert_eq!(1, 1)<span class="annotation">⦉@6</span></span><span><span class="code odd" style="--layer: 1" title="4:14-4:30: @7[0]: _0 = const ()"><span class="annotation">⦉@7</span></span></span><span class="code" style="--layer: 0">, // this is run,</span></span>
+<span class="line"><span class="code" style="--layer: 0">        2 =&gt; </span><span><span class="code even" style="--layer: 1" title="5:14-5:30: @10[0]: _0 = const ()"><span class="annotation">@10⦊</span></span></span><span class="code even" style="--layer: 2" title="5:14-5:30: @9[1]: _34 = core::panicking::AssertKind::Eq
+5:14-5:30: @9[2]: FakeRead(ForLet, _34)
+5:14-5:30: @9[5]: _36 = move _34
+5:14-5:30: @9[8]: _38 = &amp;(*_27)
+5:14-5:30: @9[9]: _37 = &amp;(*_38)
+5:14-5:30: @9[12]: _40 = &amp;(*_28)
+5:14-5:30: @9[13]: _39 = &amp;(*_40)
+5:14-5:30: @9[15]: _41 = Option::&lt;Arguments&gt;::None
+5:14-5:30: @9.Call: core::panicking::assert_failed::&lt;i32, i32&gt;(move _36, move _37, move _39, move _41) -&gt; bb17"><span class="annotation">@9⦊</span>assert_eq!(1, 1)<span class="annotation">⦉@9</span></span><span><span class="code even" style="--layer: 1" title="5:14-5:30: @10[0]: _0 = const ()"><span class="annotation">⦉@10</span></span></span><span class="code" style="--layer: 0">, // this,</span></span>
+<span class="line"><span class="code" style="--layer: 0">        3 =&gt; </span><span><span class="code odd" style="--layer: 1" title="6:14-6:30: @13[0]: _0 = const ()"><span class="annotation">@13⦊</span></span></span><span class="code even" style="--layer: 2" title="6:14-6:30: @12[1]: _54 = core::panicking::AssertKind::Eq
+6:14-6:30: @12[2]: FakeRead(ForLet, _54)
+6:14-6:30: @12[5]: _56 = move _54
+6:14-6:30: @12[8]: _58 = &amp;(*_47)
+6:14-6:30: @12[9]: _57 = &amp;(*_58)
+6:14-6:30: @12[12]: _60 = &amp;(*_48)
+6:14-6:30: @12[13]: _59 = &amp;(*_60)
+6:14-6:30: @12[15]: _61 = Option::&lt;Arguments&gt;::None
+6:14-6:30: @12.Call: core::panicking::assert_failed::&lt;i32, i32&gt;(move _56, move _57, move _59, move _61) -&gt; bb17"><span class="annotation">@12⦊</span>assert_eq!(1, 1)<span class="annotation">⦉@12</span></span><span><span class="code odd" style="--layer: 1" title="6:14-6:30: @13[0]: _0 = const ()"><span class="annotation">⦉@13</span></span></span><span class="code" style="--layer: 0">, // and this too</span></span>
+<span class="line"><span class="code" style="--layer: 0">        _ =&gt; </span><span><span class="code even" style="--layer: 1" title="7:14-7:30: @15[0]: _0 = const ()"><span class="annotation">@15⦊</span></span></span><span class="code even" style="--layer: 2" title="7:14-7:30: @14[1]: _74 = core::panicking::AssertKind::Eq
+7:14-7:30: @14[2]: FakeRead(ForLet, _74)
+7:14-7:30: @14[5]: _76 = move _74
+7:14-7:30: @14[8]: _78 = &amp;(*_67)
+7:14-7:30: @14[9]: _77 = &amp;(*_78)
+7:14-7:30: @14[12]: _80 = &amp;(*_68)
+7:14-7:30: @14[13]: _79 = &amp;(*_80)
+7:14-7:30: @14[15]: _81 = Option::&lt;Arguments&gt;::None
+7:14-7:30: @14.Call: core::panicking::assert_failed::&lt;i32, i32&gt;(move _76, move _77, move _79, move _81) -&gt; bb17"><span class="annotation">@14⦊</span>assert_eq!(1, 2)<span class="annotation">⦉@14</span></span><span><span class="code even" style="--layer: 1" title="7:14-7:30: @15[0]: _0 = const ()"><span class="annotation">⦉@15</span></span></span><span class="code" style="--layer: 0">, // however this is not</span></span>
 <span class="line"><span class="code" style="--layer: 0">    }</span></span>
-<span class="line"><span class="code" style="--layer: 0">}</span><span><span class="code odd" style="--layer: 1" title="9:2-9:2: @28.Return: return"><span class="annotation">@28⦊</span>‸<span class="annotation">⦉@28</span></span></span></span></div>
+<span class="line"><span class="code" style="--layer: 0">}</span><span><span class="code odd" style="--layer: 1" title="9:2-9:2: @16.Return: return"><span class="annotation">@16⦊</span>‸<span class="annotation">⦉@16</span></span></span></span></div>
 </body>
 </html>
make[1]: Leaving directory '/d/a/rust/rust/src/test/run-make-fulldeps/coverage-spanview'
------------------------------------------
stderr:
------------------------------------------
------------------------------------------
make[1]: *** [Makefile:50: doctest_crate] Error 1
------------------------------------------




failures:
    [run-make] run-make-fulldeps\coverage-spanview

test result: FAILED. 181 passed; 1 failed; 36 ignored; 0 measured; 0 filtered out; finished in 54.39s



command did not execute successfully: "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0-tools-bin\\compiletest.exe" "--compile-lib-path" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2\\bin" "--run-lib-path" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "--rustc-path" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2\\bin\\rustc.exe" "--rustdoc-path" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage2\\bin\\rustdoc.exe" "--rust-demangler-path" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage0-tools-bin\\rust-demangler.exe" "--src-base" "D:\\a\\rust\\rust\\src/test\\run-make-fulldeps" "--build-base" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make-fulldeps" "--stage-id" "stage2-x86_64-pc-windows-msvc" "--suite" "run-make-fulldeps" "--mode" "run-make" "--target" "x86_64-pc-windows-msvc" "--host" "x86_64-pc-windows-msvc" "--llvm-filecheck" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\llvm\\build\\bin\\FileCheck.exe" "--nodejs" "C:\\Program Files\\nodejs\\node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\native\\rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\native\\rust-test-helpers" "--docck-python" "C:\\hostedtoolcache\\windows\\Python\\3.9.1\\x64\\python3.exe" "--lldb-python" "C:\\hostedtoolcache\\windows\\Python\\3.9.1\\x64\\python3.exe" "--gdb" "C:\\msys64\\usr\\bin\\gdb" "--llvm-version" "11.0.1-rust-1.51.0-nightly" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions frontendopenmp fuzzmutate globalisel gtest gtest_main hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target testingsupport textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xray" "--llvm-bin-dir" "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\llvm\\build\\bin" "--cc" "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX64\\x64\\cl.exe" "--cxx" "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX64\\x64\\cl.exe" "--cflags" "-nologo -MT -Brepro" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"



make: *** [Makefile:72: ci-subset-1] Error 1
failed to run: D:\a\rust\rust\build\bootstrap\debug\bootstrap test --stage 2 --exclude src/test/ui --exclude src/tools/linkchecker
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1836704423)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1836704423","builder_name":"checks-actions"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #78052) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

tmiasko commented:
Looking at last bors failure, it is likely that you will have to set `profiler = true` in `config.toml` and additionally bless `run-make-fulldeps` tests with `./x.py test src/test/run-make-fulldeps/ --bless`.

a1phyr committed:
Fix run-make-fulldeps test

a1phyr commented:
Thanks ! I didn't know about the `profiler` config, so I thought the issue only happened on `msvc` target.

a1phyr commented:
@m-ou-se I don't think I can retry myself, could you do it please ?

m-ou-se commented:
Oh sorry, missed your previous message.

@bors r+ rollup=maybe

bors commented:
:pushpin: Commit 7333759502f9103dfe9f145e0b01ce547c76f165 has been approved by `m-ou-se`

<!-- @bors r=m-ou-se 7333759502f9103dfe9f145e0b01ce547c76f165 -->
<!-- homu: {"type":"Approved","sha":"7333759502f9103dfe9f145e0b01ce547c76f165","approver":"m-ou-se"} -->

panstromek commented:
Note: this should probably be `rollup=never`, it affects perf

jyn514 commented:
@bors rollup=never

jyn514 commented:
also @m-ou-se `rollup=maybe` is not the same as `rollup=iffy`, it's the same as not adding `rollup` at all: https://forge.rust-lang.org/compiler/reviews.html#rollups

m-ou-se commented:
Oh I'm aware of `rollup=maybe` undoing the rollup command, but I didn't mean to apply it to this PR though. Looks like I got it mixed up with another PR I was looking at while trying to multitask, oops. Thanks for catching my mistake, @jyn514.

bors commented:
:hourglass: Testing commit 7333759502f9103dfe9f145e0b01ce547c76f165 with merge ed58a2b03b6284b070fae2349898b16df98b7765...
<!-- homu: {"type":"BuildStarted","head_sha":"7333759502f9103dfe9f145e0b01ce547c76f165","merge_sha":"ed58a2b03b6284b070fae2349898b16df98b7765"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1945238919)
Approved by: m-ou-se
Pushing ed58a2b03b6284b070fae2349898b16df98b7765 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"m-ou-se","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1945238919"},"merge_sha":"ed58a2b03b6284b070fae2349898b16df98b7765"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1945239851) [(plain)](https://github.com/rust-lang-ci/rust/commit/ed58a2b03b6284b070fae2349898b16df98b7765/checks/1945239851/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at ed58a2b03 Auto merge of #79100 - a1phyr:better_assert_eq, r=m-ou-se
##[group]Run src/ci/publish_toolstate.sh
src/ci/publish_toolstate.sh
env:
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  DEPLOY_BUCKET: rust-lang-ci2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
  CACHE_DOMAIN: ci-caches.rust-lang.org
  TOOLSTATE_REPO_ACCESS_TOKEN: ***
##[endgroup]
Cloning into 'rust-toolstate'...
/home/runner/work/rust/rust/src/tools/publish_toolstate.py:121: DeprecationWarning: 'U' mode is deprecated
📣 Toolstate changed by rust-lang/rust#79100!
  with open(path, 'rU') as f:

Tested on commit rust-lang/rust@ed58a2b03b6284b070fae2349898b16df98b7765.
Direct link to PR: <https://github.com/rust-lang/rust/pull/79100>

🎉 miri on windows: test-fail → test-pass (cc @oli-obk @eddyb @RalfJung).
🎉 miri on linux: test-fail → test-pass (cc @oli-obk @eddyb @RalfJung).
Traceback (most recent call last):
Traceback (most recent call last):
  File "/home/runner/work/rust/rust/src/tools/publish_toolstate.py", line 338, in <module>
    response = urllib2.urlopen(urllib2.Request(
  File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 522, in open
    req = meth(req)
  File "/usr/lib/python3.8/urllib/request.py", line 1281, in do_request_
    raise TypeError(msg)
TypeError: POST data should be bytes, an iterable of bytes, or a file object. It cannot be of type str.
##[error]Process completed with exit code 1.
```

</details>

hyd-dev commented on library/core/src/panicking.rs:
```diff
@@ -91,3 +91,54 @@ pub fn panic_fmt(fmt: fmt::Arguments<'_>) -> ! {
+            Some(args) => panic!(
+                r#"assertion failed: `(left {} right)`
+  left: `{:?}`,
+ right: `{:?}: {}`"#,
```
Is the change that moves `: {}` inside the <code>\`</code> quotes intentional?

The panic message now looks suboptimal to me: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0cbe6c5669d0b7458c6d9f67a3ea7518

##### Nightly
```
 right: `3: 1 + 1 should definitely be 3`', src/main.rs:2:5
```

##### Stable
```
 right: `3`: 1 + 1 should definitely be 3', src/main.rs:2:5
```

m-ou-se replied:
@hyd-dev oh, good catch! Can you send that patch as a PR? (Feel free to assign me with `r? @m-ou-se`.)
